core/embed/sys/dbg/syslog.c

```diff
@@ -0,0 +1,293 @@
+#define SYSLOG_MAX_FILTER_LEN 128
+
+#define ESC_COLOR_NORMAL "\e[0m"
+#define ESC_COLOR_SOURCE "\e[35m"
+#define ESC_COLOR_ERR "\e[31m"
+#define ESC_COLOR_WARN "\e[33m"
+#define ESC_COLOR_INF "\e[36m"
+#define ESC_COLOR_DBG "\e[32m"
+
+typedef struct {
+  // Current filter string
+  char filter[SYSLOG_MAX_FILTER_LEN];
+  // Not ended record
+  bool eol_needed;
+} syslog_t;
+
+static syslog_t g_syslog;
+
+static bool syslog_filter_match(const log_source_t* source, log_level_t level) {
+  syslog_t* syslog = &g_syslog;
+
+  const char* p = syslog->filter;
+
+  // Start with inclusion of everything if the filter is empty
+  // or starts with '-';
+  bool included = (*p == '-' || *p == '\0');
+
+  while (*p != '\0') {
+    // Parse operation
+    char op = *p++;
+    if (op != '-' && op != '+') {
+      // Error in filter format
+      break;
+    }
+
+    // Parse log level threshold
+    log_level_t threshold = op == '-' ? LOG_LEVEL_ERR : LOG_LEVEL_DBG;
+    if (*p >= '1' && *p <= '4') {
+      threshold = LOG_LEVEL_OFF + (uint8_t)(*p - '0');
+      p++;
+    }
+
+    const char* s = source->name;
+    const char* s_end = s + source->name_len;
+
+    // Parse module name
+    while (*p != '\0' && s < s_end && *p == *s) {
+      p++;
+      s++;
+    }
+
+    // Wildcard at the end?
+    if (*p == '*') {
+      p++;
+      s = s_end;
+    } else {
+      // Skip remaining characters in module name
+      while (*p != '\0' && *p != '-' && *p != '+') {
+        p++;
+      }
+    }
+
+    // Skip spaces
+    while (*p == ' ') {
+      p++;
+    }
+
+    // Module name matched?
+    if (s == s_end && (*p == '+' || *p == '-' || *p == '\0')) {
+      if (op == '-' && level >= threshold) {
+        included = false;
+      } else if (level <= threshold) {
+        included = true;
+      }
+    }
+  }
+
+  return included;
+}
+
+static const char* log_level_str(log_level_t level) {
+  switch (level) {
+    case LOG_LEVEL_ERR:
+      return ESC_COLOR_ERR "ERR" ESC_COLOR_NORMAL;
+    case LOG_LEVEL_WARN:
+      return ESC_COLOR_WARN "WRN" ESC_COLOR_NORMAL;
+    case LOG_LEVEL_INF:
+      return ESC_COLOR_INF "INF" ESC_COLOR_NORMAL;
+    case LOG_LEVEL_DBG:
+      return ESC_COLOR_DBG "DBG" ESC_COLOR_NORMAL;
+    default:
+      return "UNK";
+  }
+}
+
+#ifndef TREZOR_EMULATOR
+static inline bool in_thread_context(void) {
+  uint32_t ipsr = __get_IPSR();
+  return (ipsr == 0 || ipsr == 11);  // Thread mode or SVCall
+}
+#endif
+
+bool syslog_start_record(const log_source_t* source, log_level_t level) {
+  syslog_t* syslog = &g_syslog;
+
+  if (syslog_filter_match(source, level)) {
+    // Prepare a record header
+    uint32_t ticks = systick_ms();
+    uint32_t seconds = ticks / 1000;
+    uint32_t msec = ticks % 1000;
+    const char* level_str = log_level_str(level);
+
+#ifndef TREZOR_EMULATOR
+    irq_key_t irq_key = irq_lock();
+#endif
+
+    const char* eol = syslog->eol_needed ? EOL_STRING : "";
+    syslog->eol_needed = true;
```

Comment:
Maybe we can simplify the logic by unconditionally prefixing `syslog_start_record()` with an EOL?

Reply:
Sounds good - let's keep the current solution.